L85-core-binder-transport.md

```diff
@@ -0,0 +1,206 @@
+# Native cross-process communication on Android using BinderChannel
+
+*   Author(s): [Ming-Chuan Lin](https://github.com/sifmelcara),
+    [Ta-Wei Tu](https://github.com/taweitu),
+    [Nanping Jiang](https://github.com/napathome)
+*   Approver: TBD
+*   Status: Draft
+*   Implemented in: gRPC-core, C++
+*   Last updated: October 5, 2021
+*   Discussion at: https://groups.google.com/g/grpc-io/c/n8Qhlfwd8tE
+
+## Abstract
+
+Implements [BinderChannel protocol](proposal L73) in gRPC-core, and provide C++
+interface for Android native code.
+
+## Background
+
+Secure cross-process gRPC communication on Android was made possible by
+[BinderChannel protocol]. However, it is only supported in gRPC Java.
+
+For various reasons (performance, code reuse, etc.) people sometimes have to
+write their code in C++. It would be nice if BinderChannel servers and clients
+are also supported in C++, as this would increase the interoperability between
+Java and native code on Android.
+
+Although it is possible to use [JNI] to utilize gRPC Java's implementation of
+[BinderChannel protocol], it incurs significant overhead when JNI call is
+involved in every RPC call. Also [JNI] could make the codebase harder to
+maintain.
+
+## Proposal
+
+We propose to implement [BinderChannel protocol] as a new transport in
+gRPC-Core, and expose C++ APIs along with some Java helper classes to support
+use scenario in Android native code.
+
+[NdkBinder] API will be used to implement the [wireformat] specified by the
+protocol.
+
+[JNI] and Java will be used by the transport implementation to set up the
+connection (since APIs for setting up connections are only available in Java).
+After the connection is established, the rest can be done in pure C++.
+
+### Client side API
+
+Since we need more than a simple URI string to create the channel, we will have
+separate APIs for creating binder transport channels.
+
+```cpp
+std::shared_ptr<grpc::Channel> CreateBinderChannel(
```

Comment:
Do you actually need a special channel creation function?  Could you instead pass `jni_env` and `context` via channel args, `package_name` and `class_name` via the URI, and `security_policy` via a new credential type?

Reply:
`jni_env` and `context` is mandatory for establishing the connection and they are pointers to special JVM structure. I'm not sure about the convention and assumed that `channel_args` is mainly for specifying configurations and might not be suitable for this use case. `jni_env` and `context` are non-optional and they don't feel like "channel arguments" since they are basically only used for establishing the connection and is not related to `grpc::Channel` logic.

For `package_name` and `class_name` it is possible to encode them into URI but it is simpler to not include them, since 1. `package_name` and `class_name` are also only used single-time when establishing the connection (in Java). 2. `package_name` and `class_name` is not the only way to establish the connection. For example, in the future we might support creating channel from a pointer to binder, in that case it is similar to `grpc_insecure_channel_create_from_fd` and URI does not make too much sense for that case.

For `security_policy` I think it cannot be easily implemented with gRPC's credential design (I might be wrong). gRPC credential is about writing/reading some kind of token/signature in wire data. However our security policies are using system calls to verify if the connection is allowed and has nothing to do with wire data or headers.

So in the end I decided that it is easier (for users and for us) to introduce a separate interfaces instead of squeezing everything into existing `CreateChannel` interface.